File: src/java/jcppsocket/src/socket/SyncClient.java
package socket;

import java.io.IOException;
import java.net.Socket;
import java.net.UnknownHostException;

import msgs.SocketMsg;

public class SyncClient {
	private String serverName;
	private int    serverPort;
	private Session session = null;

	public SyncClient(final String serverName, 
			final int	   serverPort) {
		this.serverName = serverName;
		this.serverPort = serverPort;
	}

	public boolean connect() {
		if(session == null) {
			try {
				Socket socket = new Socket(serverName, serverPort);
				session = new Session(socket, 42, 23);
			} catch (UnknownHostException e) {
				return false;
			} catch (IOException e) {
				return false;
			}
		}

		return true;
	}

	public boolean disconnect() {
		if(session == null) {
			return false;
		}


		try {
			session.close();
		} catch (IOException e) {
			e.printStackTrace();
			session = null;
			return false;
		}

		session = null;
		return true;
	}

	public SocketMsg query(SocketMsg out) {
		if(session == null) {
			return new msgs.Error("Socket not connected!");
		}

		try {
			return session.query(out);
		} catch (IOException e) {
			return new msgs.Error(e.toString());
		}
	}

    public SocketMsg read() {
        if (session == null) {
            return new msgs.Error("Socket not connected!");
        }

        try {
            return session.read();
        } catch (IOException e) {
            return new msgs.Error(e.toString());
        }
    }

    public boolean write(SocketMsg msg) {
        if (session == null) {
            return false;
        }

        try {
            session.write(msg);
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

	public boolean isConnected() {
		if(session == null) {
			return false;
		}

		return true;
	}
}

File: src/cpp/sync_client.cpp
/// HEADER
#include <utils_jcppsocket/cpp/sync_client.h>

/// PROJECT
#include <utils_jcppsocket/cpp/serialize.hpp>

using namespace utils_jcppsocket;
using namespace io;

SyncClient::SyncClient(const std::string &server,
               const int port) :
    server_name_(server),
    server_port_(port),
    connected_(false)
{
}

SyncClient::~SyncClient()
{
    if(connected_)
        disconnect();
}

bool SyncClient::connect()
{
    if(connected_) {
        std::cerr << "Warning: Socket is already connected!" << std::endl;
        return connected_;
    }

    try {
        getClientSession<23,42>(server_name_, server_port_, session_);
        connected_ = true;
    } catch (const std::exception &e) {
        std::cerr << e.what() << std::endl;
        connected_ = false;
        session_.reset();
    }

    SocketMsg::Ptr response;
    session_->read(response);

    ErrorMsg::Ptr err = std::dynamic_pointer_cast<ErrorMsg>(response);

    if(err.get() != nullptr) {
        std::cerr << err->get() << std::endl;
        disconnect();
        connected_ = false;
    }

    return connected_;
}

bool SyncClient::disconnect()
{
    if(!connected_) {
        std::cerr << "Warning: Socket is not connected!" << std::endl;
        return true;
    }

    try {
        session_->close();
        connected_ = false;
    } catch(std::exception &e) {
        std::cerr << e.what() << std::endl;
        return false;
    }

    return true;
}

bool SyncClient::query(SocketMsg::Ptr &request, SocketMsg::Ptr &response)
{
    if(!connected_) {
        std::cerr << "Error: Socket is not connected!" << std::endl;
        return false;
    }

    try {
        session_->query(request, response);
    } catch (const std::exception &e) {
        std::cerr << e.what() << std::endl;
        return false;
    }

    return true;

}

bool SyncClient::write(SocketMsg::Ptr &request)
{
    if(!connected_) {
        std::cerr << "Error: Socket is not connected!" << std::endl;
        return false;
    }

    try {
        session_->write(request);
    } catch( const std::exception &e) {
        std::cerr << e.what() << std::endl;
        return false;
    }

    return true;
}

bool SyncClient::read(SocketMsg::Ptr &response)
{
    if(!connected_) {
        std::cerr << "Error: Socket is not connected!" << std::endl;
        return false;
    }

    try {
        session_->read(response);
    } catch( const std::exception &e) {
        std::cerr << e.what() << std::endl;
        return false;
    }

    return true;
}

bool SyncClient::isConnected() const
{
    return connected_;
}
